src/realmd/AuthSocket.cpp

```diff
@@ -636,8 +650,17 @@ bool AuthSocket::_HandleLogonProof()
     BigNumber M;
     M.SetBinary(sha.GetDigest(), 20);
 
+    ///- Check PIN data is correct
+    bool pinResult = true;
+
+    if(securityFlags & SECURITY_FLAG_PIN && !lp.securityFlags & SECURITY_FLAG_PIN)
```

Comment:
This looks quite suspicious. Remember that in C++ the logical not (operator!) has priority over the bitwise and (operator&), so you're effectively checking first if ```lp.securityFlags``` is non zero and then you're doing a bitwise and of that bool with SECURITY_FLAG_PIN. I think this needs to be ```!(lp.securityFlags & SECURITY_FLAG_PIN)``` or equivalent like ```~lp.securityFlags & SECURITY_FLAG_PIN```, but the first is probably easier to understand.

Reply:
The code is refactored in a later commit and looks like this now; https://github.com/cmangos/mangos-classic/pull/377/files#diff-d7bd4c0307846a89da8bec996236180fR648 - It is no longer using a logical not operator on the flags.